noria-server/mir/src/optimize.rs

```diff
@@ -57,64 +58,56 @@ fn find_and_merge_filter_aggregates(q: &mut MirQuery) -> Vec<MirNodeRef> {
+                    if child.ancestors.len() == 1 {
+                        candidate = true;
+                    }
                 }
             },
             MirNodeType::Aggregation { ref on, .. }  => {
-                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // if the child is a filter and it has exactly one parent,
                 // then this is a candidate
-                if n.borrow().children.len() == 1 {
-                    let temp = n.borrow();
-                    let child = temp.children.first().unwrap().borrow();
-                    match child.inner {
-                        MirNodeType::Filter { ref conditions } => {
-                            if child.ancestors.len() == 1 {
-                                candidate = true;
-
-                                // But wait -- need to check if the filter is on the aggregation result
-                                use nom_sql::FunctionExpression::{Count, Sum};
-                                for (i, col) in child.columns.iter().enumerate() {
-                                    match col.function.clone() {
-                                        Some(func_expr) => {
-                                            match *func_expr {
-                                                 Count(col, _)
-                                                 | Sum(col, _) => {
-                                                    if col.name == on.name {
-                                                        // this column may be the aggregation result
-                                                        // so if we're filtering on it, we're not a candidate
-                                                        for (j, _cond) in conditions {
-                                                            if *j == i {
-                                                                candidate = false;
-                                                            }
-                                                        }
-                                                    }
-                                                },
-                                                _ => {},
-                                            }
-                                        },
-                                        None => {},
+                if let MirNodeType::Filter { ref conditions } = child.inner {
+                    if child.ancestors.len() != 1 {
+                        continue;
+                    }
+                    candidate = true;
+
+                    // But wait -- need to check if the filter is on the aggregation result
+                    use nom_sql::FunctionExpression::{Count, Sum};
+                    for (i, col) in child.columns.iter().enumerate() {
+                        if let Some(func_expr) = col.function.clone() {
```

Comment:
I don't _think_ this needs to `.clone`(). Can't it just be:
```rust
match col.function {
  Some(Count(ref col, _)) | Some(Sum(ref col, _)) if col.name == on.name => {
```

Reply:
This seems to fail because the FunctionExpression is boxed. Is there a way to write something like Some(Box(Count(...)))? I don't know much about boxes.

Reply:
Ah, yes, that's a little awkward. Easiest way is
```rust
match col.function.as_deref() {
```